bindings/java/src/main/com/apple/foundationdb/ReadTransaction.java

```diff
@@ -466,8 +460,50 @@ AsyncIterable<KeyValue> getRange(Range range,
 	 * @return a handle to access the results of the asynchronous call
 	 */
 	AsyncIterable<MappedKeyValue> getMappedRange(KeySelector begin, KeySelector end, byte[] mapper, int limit,
-	                                             int matchIndex, boolean reverse, StreamingMode mode);
-
+	                                             boolean reverse, StreamingMode mode);
+	/**
+	 * WARNING: This feature is considered experimental at this time. It is only allowed when using snapshot isolation
+	 * AND disabling read-your-writes.
+	 *
+	 * @see KeySelector
+	 * @see AsyncIterator
+	 *
+	 * @param begin the beginning of the range (inclusive)
+	 * @param end the end of the range (exclusive)
+	 * @param mapper defines how to map a key-value pair (one of the key-value pairs got
+	 *  from the first range query) to a GetRange (or GetValue) request.
+	 *  more details: https://github.com/apple/foundationdb/wiki/Everything-about-GetMappedRange
+	 * @param limit the maximum number of results to return. Limits results to the
+	 *  <i>first</i> keys in the range. Pass {@link #ROW_LIMIT_UNLIMITED} if this query
+	 *  should not limit the number of results. If {@code reverse} is {@code true} rows
+	 *  will be limited starting at the end of the range.
+	 * @param reverse return results starting at the end of the range in reverse order.
+	 *  Reading ranges in reverse is supported natively by the database and should
+	 *  have minimal extra cost.
+	 * @param mode provide a hint about how the results are to be used. This
+	 *  can provide speed improvements or efficiency gains based on the caller's
+	 *  knowledge of the upcoming access pattern.
+	 * @param matchIndex the mode to return index entries based on whether their
+	 *  corresponding records are present, examples:
+	 *     {@link FDBTransaction#MATCH_INDEX_ALL}
+	 *     {@link FDBTransaction#MATCH_INDEX_NONE}
+	 *     {@link FDBTransaction#MATCH_INDEX_MATCHED_ONLY}
+	 *     {@link FDBTransaction#MATCH_INDEX_UNMATCHED_ONLY}
+	 * <p>
+	 *     When converting the result of this query to a list using {@link AsyncIterable#asList()} with the {@code
+	 * ITERATOR} streaming mode, the query is automatically modified to fetch results in larger batches. This is done
+	 * because it is known in advance that the {@link AsyncIterable#asList()} function will fetch all results in the
+	 * range. If a limit is specified, the {@code EXACT} streaming mode will be used, and otherwise it will use {@code
+	 * WANT_ALL}.
+	 *
+	 *     To achieve comparable performance when iterating over an entire range without using {@link
+	 * AsyncIterable#asList()}, the same streaming mode would need to be used.
+	 * </p>
+	 * @return a handle to access the results of the asynchronous call
+	 */
+	AsyncIterable<MappedKeyValue> getMappedRangeOptionalIndex(KeySelector begin, KeySelector end, byte[] mapper,
```

Comment:
it is doable, however in this case, do we still need to support the 7.2 java + 7.2 primary C + 7.1 external C + 7.1 server at all? Because 7.2 java side has to adapt the change about the new parameter, thus it no longer has a 7.1 fashion java API, so there is no reason to support this case because it would not compile anyway if 7.2 java tries to do 7.1 fashion?

Reply:
[here](https://github.com/apple/foundationdb/blob/main/fdbclient/include/fdbclient/ISingleThreadTransaction.h#L73) we have default value for the methods, and it will call [C API here](https://github.com/apple/foundationdb/blob/main/fdbclient/MultiVersionTransaction.actor.cpp#L166), so we have to differentiate it by having a different name, or we cannot add this new parameter at the end, instead we need to insert it in the middle without a default value.